server/src/main/java/io/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java

```diff
@@ -0,0 +1,423 @@
+        AppenderatorDriverMetadata.class
+    );
+
+    log.info("Restored metadata[%s].", metadata);
+
+    if (metadata != null) {
+      synchronized (segments) {
+        final Map<String, String> lastSegmentIds = metadata.getLastSegmentIds();
+        Preconditions.checkState(
+            metadata.getSegments().keySet().equals(lastSegmentIds.keySet()),
+            "Sequences for segment states and last segment IDs are not same"
+        );
+
+        final Map<String, SegmentsForSequenceBuilder> builders = new TreeMap<>();
+
+        for (Entry<String, List<SegmentWithState>> entry : metadata.getSegments().entrySet()) {
+          final String sequenceName = entry.getKey();
+          final SegmentsForSequenceBuilder builder = new SegmentsForSequenceBuilder(lastSegmentIds.get(sequenceName));
+          builders.put(sequenceName, builder);
+          entry.getValue().forEach(builder::add);
+        }
+
+        builders.forEach((sequence, builder) -> segments.put(sequence, builder.build()));
+      }
+
+      return metadata.getCallerMetadata();
+    } else {
+      return null;
+    }
+  }
+
+  public AppenderatorDriverAddResult add(
+      InputRow row,
+      String sequenceName,
+      final Supplier<Committer> committerSupplier
+  ) throws IOException
+  {
+    return append(row, sequenceName, committerSupplier, false, true);
+  }
+
+  /**
+   * Add a row. Must not be called concurrently from multiple threads.
+   *
+   * @param row                      the row to add
+   * @param sequenceName             sequenceName for this row's segment
+   * @param committerSupplier        supplier of a committer associated with all data that has been added, including this row
+   *                                 if {@param allowIncrementalPersists} is set to false then this will not be used
+   * @param skipSegmentLineageCheck  if true, perform lineage validation using previousSegmentId for this sequence.
+   *                                 Should be set to false if replica tasks would index events in same order
+   * @param allowIncrementalPersists whether to allow persist to happen when maxRowsInMemory or intermediate persist period
+   *                                 threshold is hit
+   *
+   * @return {@link AppenderatorDriverAddResult}
+   *
+   * @throws IOException if there is an I/O error while allocating or writing to a segment
+   */
+  public AppenderatorDriverAddResult add(
+      final InputRow row,
+      final String sequenceName,
+      final Supplier<Committer> committerSupplier,
+      final boolean skipSegmentLineageCheck,
+      final boolean allowIncrementalPersists
+  ) throws IOException
+  {
+    return append(row, sequenceName, committerSupplier, skipSegmentLineageCheck, allowIncrementalPersists);
+  }
+
+  /**
+   * Move a set of identifiers out from "active", making way for newer segments.
+   */
+  public void moveSegmentOut(final String sequenceName, final List<SegmentIdentifier> identifiers)
```

Comment:
can you comment here saying this method only exists to support KafkaIndexTask.runLegacy(..) and should be removed with that.

Reply:
Thanks. Added.